Server crashes while after build-on-save zls tries to processMessageReportError

### Zig Version

0.16.0-dev.1657+985a3565c

### ZLS Version

0.16.0-dev.77+d84c451f

### Client / Code Editor / Extensions

Visual Studio Code (1.107.0)

### Steps to Reproduce and Observed Behavior

Behavior (after saving):
- No changes in editor
- After 5 saves VSCode saying "Server will not be restarted"

Steps to reproduce:
1. Download ZLS official extension in VSCode
2. Open any project build with Zig
3. Try to save file
4. In `OUTPUT` section will be logs with crash

### Expected Behavior

Working error highlighting

### Log Output

```shell
info  ( main ): Starting ZLS      0.16.0-dev.77+d84c451f @ '/home/ubuntu/.config/Code/User/globalStorage/ziglang.vscode-zig/zls/aarch64-linux-0.16.0-dev.77+d84c451f/zls'
info  ( main ): Log File:         /home/ubuntu/.cache/zls/zls.log (info)
info  (server): Client Info:      Visual Studio Code (1.107.0)
info  (server): added Workspace Folder: file:///home/ubuntu/Projects/Zig/ziglings
info  (server): Set config option 'enable_build_on_save' to true
info  (server): Set config option 'semantic_tokens' to "partial"
info  (server): Set config option 'inlay_hints_hide_redundant_param_names' to true
info  (server): Set config option 'warn_style' to true
info  (server): Set config option 'builtin_path' to "/home/ubuntu/.cache/zls/builtin.zig"
info  (server): Set config option 'zig_lib_path' to "/home/ubuntu/.local/zig/lib"
info  (server): Set config option 'zig_exe_path' to "/home/ubuntu/.local/zig/zig"
info  (server): Set config option 'build_runner_path' to "/home/ubuntu/.cache/zls/build_runner/01e80f3ae08c8c7bd679a68ca70be783/build_runner.zig"
info  (server): Set config option 'global_cache_path' to "/home/ubuntu/.cache/zls"
info  (server): trying to start Build-On-Save for 'file:///home/ubuntu/Projects/Zig/ziglings'
warn  (server): Build-On-Save cannot run in watch mode because it is not supported by Linux '5.10.0' (requires at least 6.5.0)
warn  ( diag ): zig build runner exited with non-zero status: 2
stderr:
         _       _ _
     ___(_) __ _| (_)_ __   __ _ ___
    |_  | |/ _' | | | '_ \ / _' / __|
     / /| | (_| | | | | | | (_| \__ \
    /___|_|\__, |_|_|_| |_|\__, |___/
           |___/           |___/

    "Look out! Broken programs below!"

Compiling: 022_errors2.zig
error: 1 compilation errors
exercises/022_errors2.zig:22:24: error: expected type expression, found '='
    var my_number: ??? = 5;
                       ^

Ziglings hint: Get the error union type right to allow this to compile.
Edit exercises/022_errors2.zig and run 'zig build' again.
```

Additional info in OUTPUT section in VSCode:
```
2025-12-26 20:15:18.100 [error] thread 610878 panic: attempt to use null value
2025-12-26 20:15:18.261 [error] /home/runner/work/zls/zls/src/Server.zig:1848:76: 0x10b138b in processMessage (zls)
2025-12-26 20:15:18.263 [error] /home/runner/work/zls/zls/src/Server.zig:1859:30: 0x109333f in processMessageReportError (zls)
2025-12-26 20:15:18.263 [error] /home/runner/work/zls/zls/src/Server.zig:1746:45: 0x1098f77 in loop (zls)
2025-12-26 20:15:18.263 [error] /home/runner/work/_temp/fcfda699-096a-4dba-a700-8dc07c787d9e/zig-x86_64-linux-0.16.0-dev.1657+985a3565c/lib/std/start.zig:714:37: 0x1093217 in callMain (zls)
2025-12-26 20:15:18.281 [error] Server process exited with signal SIGABRT.
2025-12-26 20:15:18.287 [error] The ZLS language server server crashed 5 times in the last 3 minutes. The server will not be restarted. See the output for more information.
```

--- a/src/features/diagnostics.zig
+++ b/src/features/diagnostics.zig
@@ -475,11 +475,16 @@ pub fn getErrorBundleFromStderr(
 pub const BuildOnSave = struct {
     io: std.Io,
     allocator: std.mem.Allocator,
-    child_process: *std.process.Child,
     worker: std.Io.Future(void),
+    worker_state: *WorkerState,
 
     const shared = @import("../build_runner/shared.zig");
     const ServerToClient = shared.ServerToClient;
+
+    const WorkerState = struct {
+        mutex: std.Io.Mutex,
+        child_process: std.process.Child,
+    };
 
     pub const InitOptions = struct {
         io: std.Io,
@@ -495,9 +500,6 @@ pub const BuildOnSave = struct {
     };
 
     pub fn init(options: InitOptions) error{ OutOfMemory, ConcurrencyUnavailable }!?BuildOnSave {
-        const child_process = try options.allocator.create(std.process.Child);
-        errdefer options.allocator.destroy(child_process);
-
         const base_args: []const []const u8 = &.{
             options.zig_exe_path,
             "build",
@@ -517,14 +519,13 @@ pub const BuildOnSave = struct {
         if (options.check_step_only) argv.appendAssumeCapacity("--check-only");
         argv.appendSliceAssumeCapacity(options.build_on_save_args);
 
-        child_process.* = .init(argv.items, options.allocator);
+        var child_process: std.process.Child = .init(argv.items, options.allocator);
         child_process.stdin_behavior = .Pipe;
         child_process.stdout_behavior = .Pipe;
         child_process.stderr_behavior = .Pipe;
         child_process.cwd = options.workspace_path;
 
         child_process.spawn(options.io) catch |err| {
-            options.allocator.destroy(child_process);
             log.err("failed to spawn zig build process: {}", .{err});
             return null;
         };
@@ -536,7 +537,7 @@ pub const BuildOnSave = struct {
             _ = terminateChildProcessReportError(
                 options.io,
                 options.allocator,
-                child_process,
+                &child_process,
                 "zig build runner",
                 .kill,
             );
@@ -545,10 +546,18 @@ pub const BuildOnSave = struct {
         const duped_workspace_path = try options.allocator.dupe(u8, options.workspace_path);
         errdefer options.allocator.free(duped_workspace_path);
 
+        const worker_state = try options.allocator.create(WorkerState);
+        errdefer options.allocator.destroy(worker_state);
+
+        worker_state.* = .{
+            .mutex = .init,
+            .child_process = child_process,
+        };
+
         const worker = try options.io.concurrent(loop, .{
             options.io,
             options.allocator,
-            child_process,
+            worker_state,
             options.collection,
             duped_workspace_path,
         });
@@ -557,38 +566,45 @@ pub const BuildOnSave = struct {
         return .{
             .io = options.io,
             .allocator = options.allocator,
-            .child_process = child_process,
             .worker = worker,
+            .worker_state = worker_state,
         };
     }
 
     pub fn deinit(self: *BuildOnSave) void {
         self.worker.cancel(self.io);
-        self.allocator.destroy(self.child_process);
+        self.allocator.destroy(self.worker_state);
         self.* = undefined;
     }
 
-    pub fn sendManualWatchUpdate(self: *BuildOnSave) void {
-        self.child_process.stdin.?.writeStreamingAll(self.io, "\x00") catch {};
+    pub fn sendManualWatchUpdate(build_on_save: *BuildOnSave) void {
+        build_on_save.worker_state.mutex.lockUncancelable(build_on_save.io);
+        defer build_on_save.worker_state.mutex.unlock(build_on_save.io);
+        if (build_on_save.worker_state.child_process.stdin) |stdin| {
+            stdin.writeStreamingAll(build_on_save.io, "\x00") catch {};
+        }
     }
 
     fn loop(
         io: std.Io,
         allocator: std.mem.Allocator,
-        child_process: *std.process.Child,
+        state: *WorkerState,
         collection: *DiagnosticsCollection,
         workspace_path: []const u8,
     ) void {
-        defer allocator.free(workspace_path);
-
         defer {
-            child_process.stdin.?.close(io);
-            child_process.stdin = null;
+            allocator.free(workspace_path);
+
+            state.mutex.lockUncancelable(io);
+            defer state.mutex.unlock(io);
+
+            state.child_process.stdin.?.close(io);
+            state.child_process.stdin = null;
 
             _ = terminateChildProcessReportError(
                 io,
                 allocator,
-                child_process,
+                &state.child_process,
                 "zig build runner",
                 .wait,
             );
@@ -603,19 +619,27 @@ pub const BuildOnSave = struct {
         }
 
         var read_buffer: [@sizeOf(ServerToClient.Header)]u8 = undefined;
-        var file_reader = child_process.stdout.?.reader(io, &read_buffer);
+        var file_reader = state.child_process.stdout.?.reader(io, &read_buffer);
         const reader = &file_reader.interface;
 
         while (true) {
             const header = reader.takeStruct(ServerToClient.Header, .little) catch |err| switch (err) {
                 error.ReadFailed => {
-                    log.err("failed to receive message from zig build runner: {}", .{err});
+                    if (file_reader.err.? != error.Canceled) {
+                        log.err("failed to receive message from zig build runner: {}", .{file_reader.err.?});
+                    }
                     return;
                 },
                 error.EndOfStream => break,
             };
             const body = reader.readAlloc(allocator, header.bytes_len) catch |err| switch (err) {
-                error.ReadFailed, error.EndOfStream, error.OutOfMemory => {
+                error.ReadFailed => {
+                    if (file_reader.err.? != error.Canceled) {
+                        log.err("failed to receive message from zig build runner: {}", .{file_reader.err.?});
+                    }
+                    return;
+                },
+                error.EndOfStream, error.OutOfMemory => {
                     log.err("failed to receive message from zig build runner: {}", .{err});
                     return;
                 },

